### ISSUE TITLE
Coffeescript Support

### ISSUE THREAD

**Issue by ryanpager** [SocketCluster/socketcluster#103]
I've been searching around for a good way to do this, but the long story short is =>

Is there a way to make socketCluster support coffeescript? My server is written in coffee, and integrating socket cluster has been a pain because the worker/broker files require a js (not coffee) file.

Any help on this one?

**Comment by ryanpager** [comment]
Example:

```coffeescript
opts =
  # General Settings
  appName: 'TalkFusionSS'

  # Process Settings
  port: ProcessConfig.port or 8000
  workers: ProcessConfig.maxWorkers or 1
  brokers: ProcessConfig.maxBrokers or 1
  rebootWorkerOnCrash: true
  pingTimeout: 5000
  pingInterval: 2000
  
  # Entry Point for socket events
  workerController: './worker.coffee'

  # Entry Point for socket event management
  brokerController: './broker.coffee'
```

**Comment by IngwiePhoenix** [comment]
There is no practical way.

SocketCluster uses `child_process.fork` in some places - which means, the original "intro script" is completely abandoned, leaving no room to just use CoffeeScript's `require()` "registration".

The only way this can really work is, by having SC an option that allows to prepare global scope, or even better, a function. In a similar project that I wrote exactly for problems like that and without many features which simply serves as a launcher, I did this:

```javascript
PowerHouse({
    master: function(){
        // This function runs only on the master process.
    },
    child: function(){
        // This function runs on all children.
    }
});
```

So basically, Socketcluster could serialize this `.child` method and execute it _before_ it includes the actual script. That way, you'd get a brief chance to modify global scope. That should work around CoffeeScript's global registrar.

This is all I can say for now. The only alternative you have is to manually run the registrar and then require the CoffeeScript itself...which does, even to me, seem off. Just as an example:

```javascript
SC({...controller: "myfile.js"});

// myfile.js
require("coffeescript/global");
require("myrealfile.coffee");
```

Forgive me if I got the global registrar wrong, I don't work with CoffeeScript. (: I only read about how it does it.

**Comment by IngwiePhoenix** [comment]
Oh nevermind! O.O I must've been waaaaay behind right there. Sorry for my, i think now, pointless answer. (:

**Comment by ryanpager** [comment]
So this was the solution (unless someone else tells me otherwise):

###### Broker.js
```js
require('coffee-script/register');
module.exports.run = require("./broker.coffee");
```

###### Broker.coffee

```coffeescript
module.exports = (broker) ->
  Console.debug "Broker instance started, pid #{process.pid}"
```

...do the same thing for worker as well. This will run without issues.

**Comment by IngwiePhoenix** [comment]
There should really be a way to do some "intro" code for that stuff. I just made https://github.com/IngwiePhoenix/oj-note and would like to use OJ with SC. o.o

**Comment by jondubois** [comment]
@ryanpager Your solution seems pretty good.

@IngwiePhoenix Yes, it would be nice to have a simpler way to add custom preprocessors to handle compile-to-javascript languages like coffeescript. Not sure how we could implement this - The child_process.fork() method requires a JavaScript file ultimately.

**Comment by IngwiePhoenix** [comment]
@jondubois : o.o
```javascript
var child = function(){
    require("oj-node");
    require("coffee-script/register");
}

// Serialize
var srl = child.toString();
// ... send over...
// Execute
var fn;
eval("fn = "+srl);
```

Yes, yes. Eval is evil and this is super hacky. BUT this is only one of many ways. In fact, there are people that did it beyond what I did: https://www.npmjs.com/package/funcster

It is nothing impossible, but I would advise taking the step to pick a 3rd party tool to do so. I have seen extended `JSON` implementations that simply serialized functions too.

Now, how would that be used? Serialize the "child preparation function" and send it to the respective processes along with the configuration - deserialize it there and run it, THEN require the actual module. I can already see that this stuff happens already, in a way. When launching an SC system, a process is spawned which uses a SC internal file. Using this method to spawn the clients would give SC more control over the clients, and the possibility to obtain a serialized function and execute it before the rest.

That, at least, is how I would do it. :)

**Comment by IngwiePhoenix** [comment]
You can now modify the global namespace and "prepare" the worker object. Supply `initController` to SC. This file will be able to run before any other worker - meaning, that you can configure the execution context of the process. Add CoffeeScript's register script there, and voila, you can use CoffeeScript in your Worker/Broker controller!

**Comment by jondubois** [comment]
Hopefully the new initController feature solves this: https://github.com/SocketCluster/socketcluster/issues/116
